src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,149 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            Node<T> newNode = new Node<>(null, value, null);
+            first = newNode;
+            last = newNode;
+        } else {
+            Node<T> newNode = new Node<>(last, value, null);
+            last.next = newNode;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            Node<T> current = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(null, value, current);
+            first = newNode;
+            current.prev = newNode;
+        } else {
+            checkIndex(index);
+            Node<T> current = getNodeByIndex(index);
+            Node<T> prevNode = current.prev;
+            Node<T> newNode = new Node<>(prevNode, value, current);
+            prevNode.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        T deletedValue = current.item;
+        current.item = value;
+        return deletedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        return unlink(current);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        for (int i = 0; i < size; i++) {
+            if (current.item == object || current.item != null && current.item.equals(object)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private T unlink(Node<T> current) {
+        Node<T> nextNode = current.next;
+        Node<T> prevNode = current.prev;
+        size--;
+        if (isEmpty()) {
+            first = null;
+            last = null;
+            size = 0;
+        } else if (current.equals(first)) {
```

Comment:
Did you try to remove this part of the code locally?

Reply:
Yes, Two test have NullPointerException